test/test_kernels/test_momentum_three_block_model_free_kernel.py

```diff
@@ -0,0 +1,181 @@
+    tolerance = None
+    momentum_beta = 0.7
+    indata = KernelInputType(sparse, candidate, target_rank, svd_strategy, tolerance)
+    kernel_params = Momentum3BlockAdditionalParameters(momentum_beta)
+    kernel = Momentum3BlockModelFreeKernel(indata, kernel_params)
+    return (indata, kernel_params, kernel)
+
+
+@patch(f"{PKG}.compute_loss")
+@patch(f"{PKG}.update_H")
+@patch(f"{PKG}.update_W")
+@patch(f"{PKG}.apply_momentum")
+@patch(f"{PKG}.construct_utility")
+def test_momentum_3block_model_free_kernel_step_W0H0_given_one_iter(
+    mock_construct: Mock,
+    mock_apply_momentum: Mock,
+    mock_update_W: Mock,
+    mock_update_H: Mock,
+    mock_compute_loss: Mock,
+    fixture_with_tol_W0H0_given: Fixture,
+) -> None:
+    (indata, kernel_params, kernel) = fixture_with_tol_W0H0_given
+    shape_X = indata.sparse_matrix_X.shape
+    shape_W0 = (shape_X[0], indata.target_rank)
+    shape_H0 = (indata.target_rank, shape_X[1])
+    mock_construct.return_value = np.full(shape_X, 4.0)
+    mock_apply_momentum.return_value = np.full(shape_X, 5.0)
+    mock_update_W.return_value = np.full(shape_W0, 1.0)
+    mock_update_H.return_value = np.full(shape_H0, 1.0)
+    mock_compute_loss.return_value = 3.0
+
+    kernel.step()
+    mock_construct.assert_called_once_with(
+        indata.low_rank_candidate_L, indata.sparse_matrix_X
+    )
+    mock_apply_momentum.assert_called_once_with(
+        mock_construct.return_value,
+        indata.sparse_matrix_X,
+        kernel.momentum_beta,
+    )
+    mock_update_W.assert_called_once_with(
+        kernel_params.candidate_factor_H0, mock_apply_momentum.return_value
+    )
+    mock_update_H.assert_called_once_with(
+        mock_update_W.return_value, mock_apply_momentum.return_value
+    )
+
+    mock_compute_loss.assert_called_once_with(
+        mock_apply_momentum.return_value,
+        kernel.low_rank_candidate_L,
+        LossType.FROBENIUS,
+    )
+
+
+@patch(f"{PKG}.apply_momentum")
+def test_momentum_3block_model_free_kernel_step_apply_momentum_call_count(
+    mock_apply_momentum: Mock,
+    fixture_no_tol_W0H0_not_given: Fixture,
+) -> None:
+    (indata, _, kernel) = fixture_no_tol_W0H0_not_given
+    shape_X = indata.sparse_matrix_X.shape
+
+    mock_apply_momentum.return_value = np.full(shape_X, 5.0)
+    kernel.elapsed_iterations = 0
+    kernel.step()
+    kernel.increment_elapsed()
+
+    # apply_momentum called one time in first iteration
+    assert mock_apply_momentum.call_count == 1
+
+    kernel.step()
+
+    # apply_momentum called two times in second iteration
+    assert mock_apply_momentum.call_count == 3
+
+
+def test_momentum_3block_model_free_kernel_W0H0_initialized_if_not_given(
+    fixture_no_tol_W0H0_not_given: Fixture,
+) -> None:
+    (indata, _, kernel) = fixture_no_tol_W0H0_not_given
+
+    np.testing.assert_array_almost_equal(
+        indata.low_rank_candidate_L,
+        kernel.candidate_factor_W @ kernel.candidate_factor_H,
+    )
```

Comment:
How annoying would it be to test this for the cases where only one of `W0` and `H0` is defined? Again I'm not super worried about it, and if it requires setting up two more whole fixtures or radically reworking how these are structured, it probably isn't worth it; just noting that that is an edge case which isn't covered.